### PR TITLE
[DO NOT MERGE] [autogluon][build, test] Initial PR for training containers

### PR DESCRIPTION
**Pull request by gradientsky**
*Issue #, if available:*

## PR Checklist
- [x] I've prepended PR tag with frameworks/job this applies to : [mxnet, tensorflow, pytorch] | [ei/neuron] | [build] | [test] | [benchmark] | [ec2, ecs, eks, sagemaker]
- [x] (If applicable) I've documented below the DLC image/dockerfile this relates to
- [x] (If applicable) I've documented below the tests I've run on the DLC image
- [ ] (If applicable) I've reviewed the licenses of updated and new binaries and their dependencies to make sure all licenses are on the Apache Software Foundation Third Party License Policy Category A or Category B license list.  See [https://www.apache.org/legal/resolved.html](https://www.apache.org/legal/resolved.html).
- [x] (If applicable) I've scanned the updated and new binaries to make sure they do not have vulnerabilities associated with them.

## Pytest Marker Checklist
- [x] (If applicable) I have added the marker `@pytest.mark.model("<model-type>")` to the new tests which I have added, to specify the Deep Learning model that is used in the test (use `"N/A"` if the test doesn't use a model)
- [x] (If applicable) I have added the marker `@pytest.mark.integration("<feature-being-tested>")` to the new tests which I have added, to specify the feature that will be tested
- [ ] (If applicable) I have added the marker `@pytest.mark.multinode(<integer-num-nodes>)` to the new tests which I have added, to specify the number of nodes used on a multi-node test
- [x] (If applicable) I have added the marker `@pytest.mark.processor(<"cpu"/"gpu"/"eia"/"neuron">)` to the new tests which I have added, if a test is specifically applicable to only one processor type

#### EIA/NEURON Checklist
* When creating a PR:
- [ ] I've modified `src/config/build_config.py` in my PR branch by setting `ENABLE_EI_MODE = True` or `ENABLE_NEURON_MODE = True`
* When PR is reviewed and ready to be merged:
- [ ] I've reverted the code change on the config file mentioned above
#### Benchmark Checklist
* When creating a PR:
- [ ] I've modified `src/config/test_config.py` in my PR branch by setting `ENABLE_BENCHMARK_DEV_MODE = True`
* When PR is reviewed and ready to be merged:
- [ ] I've reverted the code change on the config file mentioned above

## Reviewer Checklist
* For reviewer, before merging, please cross-check:
- [ ] I've verified the code change on the config file mentioned above has already been reverted

*Description:*
Initial check-in for autogluon framework.

*Tests run:*

*DLC image/dockerfile:*
New docker files (new framework onboarding):
* autogluon-inference:0.2-gpu-py37-cu102-ubuntu18.04
* autogluon-inference:0.2-cpu-py37-ubuntu18.04

*Additional context:*
AutoGluon - New framework onboarding
[website](https://auto.gluon.ai/stable/index.html), [github](https://github.com/awslabs/autogluon)
Licenses are still under OSS/BLL review (no red flags).

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license. I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: autogluon/training/docker/0.3.0/py3/rapids0.19/cu102/Dockerfile.gpu [hunk 1]
```diff
@@ -0,0 +1,79 @@
+ARG CUDA_VER=10.2
+ARG LINUX_VER=ubuntu18.04
+ARG PYTHON_VER=3.7
+ARG RAPIDS_VER=0.19
+ARG FROM_IMAGE=rapidsai/rapidsai-core
+
+FROM ${FROM_IMAGE}:${RAPIDS_VER}-cuda${CUDA_VER}-base-${LINUX_VER}-py${PYTHON_VER}
```
Comment (saimidu) [inline autogluon/training/docker/0.3.0/py3/rapids0.19/cu102/Dockerfile.gpu:7]: What is the OS version used by this image?

File: autogluon/training/docker/0.3.0/py3/rapids0.19/cu102/Dockerfile.gpu [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+LABEL maintainer="Amazon AI"
+
+ARG CUDA_VER
```
Comment (saimidu) [inline autogluon/training/docker/0.3.0/py3/rapids0.19/cu102/Dockerfile.gpu:10]: Why is this defined once again?

File: autogluon/training/docker/0.3.0/py3/Dockerfile.cpu [hunk 3]
```diff
@@ -0,0 +1,152 @@
...
+  && printf "Host *\n StrictHostKeyChecking no\n" >> /root/.ssh/config
+
+# This is here to make our installed version of OpenCV work.
+# https://stackoverflow.com/questions/29274638/opencv-libdc1394-error-failed-to-initialize-libdc1394
+# TODO: Should we be installing OpenCV in our image like this? Is there another way we can fix this?
+RUN ln -s /dev/null /dev/raw1394
+
+COPY deep_learning_container.py /usr/local/bin/deep_learning_container.py
+
+RUN chmod +x /usr/local/bin/deep_learning_container.py
```
[the remark this answers is not in the document]
Comment (gradientsky) [reply autogluon/training/docker/0.3.0/py3/Dockerfile.cpu:137]: removed

File: autogluon/training/docker/0.3.0/py3/rapids0.19/cu102/Dockerfile.gpu [hunk 4]
```diff
@@ -0,0 +1,79 @@
...
+  && printf "Host *\n StrictHostKeyChecking no\n" >> /root/.ssh/config
+
+# This is here to make our installed version of OpenCV work.
+# https://stackoverflow.com/questions/29274638/opencv-libdc1394-error-failed-to-initialize-libdc1394
+# TODO: Should we be installing OpenCV in our image like this? Is there another way we can fix this?
+RUN ln -s /dev/null /dev/raw1394
+
+COPY deep_learning_container.py /usr/local/bin/deep_learning_container.py
+
+RUN chmod +x /usr/local/bin/deep_learning_container.py
```
[the remark this answers is not in the document]
Comment (gradientsky) [reply autogluon/training/docker/0.3.0/py3/rapids0.19/cu102/Dockerfile.gpu:65]: removed

File: autogluon/training/docker/0.3.0/py3/Dockerfile.cpu [hunk 5]
```diff
@@ -0,0 +1,152 @@
...
+ && chmod +x /usr/local/bin/testOSSCompliance \
+ && chmod +x ${HOME_DIR}/oss_compliance/generate_oss_compliance.sh \
+ && ${HOME_DIR}/oss_compliance/generate_oss_compliance.sh ${HOME_DIR} ${PYTHON} \
+ && rm -rf ${HOME_DIR}/oss_compliance*
+
+RUN curl -o /license.txt https://autogluon.s3.us-west-2.amazonaws.com/licenses/THIRD-PARTY-LICENSES.txt
+
+CMD ["/bin/bash"]
+
+
```
[the remark this answers is not in the document]
Comment (gradientsky) [reply autogluon/training/docker/0.3.0/py3/Dockerfile.cpu:152]: fixed

File: autogluon/training/docker/0.3.0/py3/rapids0.19/cu102/Dockerfile.gpu [hunk 6]
```diff
@@ -0,0 +1,79 @@
...
+ && chmod 775 -R ${HOME_DIR}/oss_compliance/ \
+ && cp ${HOME_DIR}/oss_compliance/test/testOSSCompliance /usr/local/bin/testOSSCompliance \
+ && chmod +x /usr/local/bin/testOSSCompliance \
+ && chmod +x ${HOME_DIR}/oss_compliance/generate_oss_compliance.sh \
+ && ${HOME_DIR}/oss_compliance/generate_oss_compliance.sh ${HOME_DIR} python3 \
+ && rm -rf ${HOME_DIR}/oss_compliance*
+
+RUN curl -o /license.txt https://autogluon.s3.us-west-2.amazonaws.com/licenses/THIRD-PARTY-LICENSES.txt
+
+CMD ["/bin/bash"]
```
[the remark this answers is not in the document]
Comment (gradientsky) [reply autogluon/training/docker/0.3.0/py3/rapids0.19/cu102/Dockerfile.gpu:79]: fixed

File: autogluon/training/docker/0.3.0/py3/Dockerfile.cpu [hunk 7]
```diff
@@ -0,0 +1,152 @@
...
+WORKDIR /
+
+RUN pip install --no-cache-dir -U torch==${TORCH_VER} torchvision==${TORCH_VISION_VER} -f https://download.pytorch.org/whl/torch_stable.html \
+    mxnet==${MXNET_VER} \
+    autogluon==${AUTOGLUON_VERSION} \
+    sagemaker-mxnet-training \
+    && pip uninstall -y jupyter-client jupyter-core jupyter-packaging jupyter-server jupyter-server-proxy \
+    jupyterlab jupyterlab-nvdashboard jupyterlab-pygments jupyterlab-server jupyterlab-widgets ipywidgets \
+    ipython-genutils ipython ipykernel torchvision bokeh plotly notebook datashader
```
[the remark this answers is not in the document]
Comment (gradientsky) [reply autogluon/training/docker/0.3.0/py3/Dockerfile.cpu:113]: good point, done

File: autogluon/training/docker/0.3.0/py3/Dockerfile.cpu [hunk 8]
```diff
@@ -0,0 +1,152 @@
...
+ENV PATH=/usr/local/bin:$PATH
+# install OpenSSL
+RUN wget -q https://www.openssl.org/source/openssl-${OPENSSL_VERSION}.tar.gz \
+ && tar -xzf openssl-${OPENSSL_VERSION}.tar.gz \
+ && cd openssl-${OPENSSL_VERSION} \
+ && ./config && make -j $(nproc) && make install \
+ && ldconfig \
+ && cd .. && rm -rf openssl-* \
+ && rmdir /usr/local/ssl/certs \
+ && ln -s /etc/ssl/certs /usr/local/ssl/certs
```
[the remark this answers is not in the document]
Comment (gradientsky) [reply autogluon/training/docker/0.3.0/py3/Dockerfile.cpu:91]: removed

File: autogluon/training/docker/0.3.0/py3/Dockerfile.cpu [hunk 9]
```diff
@@ -0,0 +1,152 @@
...
+ARG TORCH_VER=1.8.0+cpu
+ARG TORCH_VISION_VER=0.9.0+cpu
+
+# Update once 0.3.0 is released
+ARG AUTOGLUON_VERSION=0.2.1b20210625
+
+# The smdebug pipeline relies for following format to perform string replace and trigger DLC pipeline for validating
+# the nightly builds. Therefore, while updating the smdebug version, please ensure that the format is not disturbed.
+ARG SMDEBUG_VERSION=0.9.4
+ARG OPENSSL_VERSION=1.1.1k
```
[the remark this answers is not in the document]
Comment (gradientsky) [reply autogluon/training/docker/0.3.0/py3/Dockerfile.cpu:17]: removed

File: autogluon/training/docker/0.3.0/py3/Dockerfile.cpu [hunk 10]
```diff
@@ -0,0 +1,152 @@
...
+    LANG=C.UTF-8 \
+    LC_ALL=C.UTF-8 \
+    SAGEMAKER_TRAINING_MODULE=sagemaker_mxnet_container.training:main \
+    DEBIAN_FRONTEND="noninteractive" \
+    TZ="UTC"
+
+# Install dependencies
+RUN apt-get update \
+    && apt-get -y install gcc \
+        build-essential \
```
[the remark this answers is not in the document]
Comment (gradientsky) [reply autogluon/training/docker/0.3.0/py3/Dockerfile.cpu:31]: updated

File: autogluon/buildspec.yml [hunk 11]
```diff
@@ -0,0 +1,49 @@
...
+    <<: *TRAINING_REPOSITORY
+    build: &AUTOGLUON_GPU_TRAINING_PY3 false
+    image_size_baseline: 12466
+    device_type: &DEVICE_TYPE gpu
+    python_version: &DOCKER_PYTHON_VERSION py3
+    tag_python_version: &TAG_PYTHON_VERSION py37
+    cuda_version: &CUDA_VERSION cu102
+    rapids_version: &RAPIDS_VERSION rapids0.19
+    os_version: &OS_VERSION ubuntu18.04
+    tag: !join [ *VERSION, "-", *DEVICE_TYPE, "-", *TAG_PYTHON_VERSION, "-", *RAPIDS_VERSION, "-", *CUDA_VERSION, "-", *OS_VERSION ]
```
Comment (SergTogul) [inline autogluon/buildspec.yml:46]: Could you remove Rapids from tag?

File: autogluon/training/docker/0.3.0/py3/Dockerfile.cpu [hunk 12]
```diff
@@ -0,0 +1,135 @@
...
+LABEL maintainer="Amazon AI"
+
+ARG PYTHON_VERSION=3.7.10
+ARG PYTHON=python3
+ARG MXNET_VER=1.7.*
+ARG TORCH_VER=1.8.0+cpu
+ARG TORCH_VISION_VER=0.9.0+cpu
+
+# Update once 0.3.0 is released
+ARG AUTOGLUON_VERSION=0.2.1b20210625
```
Comment (SergTogul) [inline autogluon/training/docker/0.3.0/py3/Dockerfile.cpu:12]: I'd suggest to put **this** file into 0.2/py3/Dockerfile.cpu directory. And create a new docker file in 0.3/py3/ as soon as you have 0.3.0 released.

File: autogluon/buildspec.yml [hunk 13]
```diff
@@ -0,0 +1,49 @@
...
+  BuildAutogluonCPUTrainPy3DockerImage:
+    <<: *TRAINING_REPOSITORY
+    build: &AUTOGLUON_CPU_TRAINING_PY3 false
+    image_size_baseline: 3512
+    device_type: &DEVICE_TYPE cpu
+    python_version: &DOCKER_PYTHON_VERSION py3
+    tag_python_version: &TAG_PYTHON_VERSION py37
+    os_version: &OS_VERSION ubuntu18.04
+    tag: !join [ *VERSION, "-", *DEVICE_TYPE, "-", *TAG_PYTHON_VERSION, "-", *OS_VERSION ]
+    docker_file: !join [ docker/, *VERSION, /, *DOCKER_PYTHON_VERSION, /Dockerfile., *DEVICE_TYPE ]
```
Comment (SergTogul) [inline autogluon/buildspec.yml:32]: Our current convention is to name directories using short versions. 0.3.0 -> 0.3 

You can check [this buildspeck](https://github.com/aws/deep-learning-containers/blob/v1.9-mx-1.8.0-tr-gpu-py37/pytorch/buildspec.yml).

File: available_images.md [hunk 14]
```diff
@@ -107,6 +107,13 @@ You can pin your version by adding the version tag to your URL as follows:
 |PyTorch 1.8.1      |inference	|No			    |GPU 		| 3.6 (py36)			|763104351884.dkr.ecr.us-east-1.amazonaws.com/pytorch-inference:1.8.1-gpu-py36-cu111-ubuntu18.04	|
 |PyTorch 1.8.1      |inference	|No				|CPU 		| 3.6 (py36)			|763104351884.dkr.ecr.us-east-1.amazonaws.com/pytorch-inference:1.8.1-cpu-py36-ubuntu18.04			|
 
+AutoGluon training containers
+===============================
+
+| Framework         |Job Type	|CPU/GPU 	|Python Version Options	|Example URL																								 |
+|-------------------|-----------|-----------|-----------------------|------------------------------------------------------------------------------------------------------------|
+|AutoGluon 0.3.0    |training   |GPU        | 3.7 (py37)            |763104351884.dkr.ecr.us-east-1.amazonaws.com/autogluon-inference:0.3.0-gpu-py37-rapids0.19-cu102-ubuntu18.04|
```
Comment (SergTogul) [inline available_images.md:115]: Could you remove rapids from here?

File: src/config/test_config.py [hunk 15]
```diff
@@ -1,6 +1,6 @@
 # Please only set it to True if you are preparing a Benchmark related PR
 # Do remember to revert it back to False before merging any PR (including Benchmark dedicated PR)
-ENABLE_BENCHMARK_DEV_MODE = False
+ENABLE_BENCHMARK_DEV_MODE = True
```
Comment (SergTogul) [inline src/config/test_config.py:3]: We don't need it now, while we're testing a functionality. You can set it back to False.

Comment (SergTogul) [reply src/config/test_config.py:3]: Plus we changed this recently. Now we have 1 [.toml file](https://github.com/aws/deep-learning-containers/blob/master/dlc_developer_config.toml) for all the settings

File: autogluon/buildspec.yml (hunk 13 above)
Comment (gradientsky) [reply autogluon/buildspec.yml:32]: Updated

File: autogluon/buildspec.yml (hunk 11 above)
Comment (gradientsky) [reply autogluon/buildspec.yml:46]: Removed

File: autogluon/training/docker/0.3.0/py3/Dockerfile.cpu (hunk 12 above)
Comment (gradientsky) [reply autogluon/training/docker/0.3.0/py3/Dockerfile.cpu:12]: Changed to 0.2.1

File: autogluon/training/docker/0.3.0/py3/rapids0.19/cu102/Dockerfile.gpu (hunk 1 above)
Comment (gradientsky) [reply autogluon/training/docker/0.3.0/py3/rapids0.19/cu102/Dockerfile.gpu:7]: ubuntu18.04

File: available_images.md (hunk 14 above)
Comment (gradientsky) [reply available_images.md:115]: removed

File: src/config/test_config.py (hunk 15 above)
Comment (gradientsky) [reply src/config/test_config.py:3]: Rolled back

**Review by SergTogul — APPROVED** [review]

File: autogluon/training/docker/0.2/py3/Dockerfile.cpu [hunk 16]
```diff
@@ -0,0 +1,137 @@
...
+RUN HOME_DIR=/root \
+ && curl -o ${HOME_DIR}/oss_compliance.zip https://aws-dlinfra-utilities.s3.amazonaws.com/oss_compliance.zip \
+ && unzip ${HOME_DIR}/oss_compliance.zip -d ${HOME_DIR}/ \
+ && cp ${HOME_DIR}/oss_compliance/test/testOSSCompliance /usr/local/bin/testOSSCompliance \
+ && chmod +x /usr/local/bin/testOSSCompliance \
+ && chmod +x ${HOME_DIR}/oss_compliance/generate_oss_compliance.sh \
+ && ${HOME_DIR}/oss_compliance/generate_oss_compliance.sh ${HOME_DIR} ${PYTHON} \
+ && rm -rf ${HOME_DIR}/oss_compliance*
+
+RUN curl -o /license.txt https://autogluon.s3.us-west-2.amazonaws.com/licenses/THIRD-PARTY-LICENSES.txt
```
Comment (tejaschumbalkar) [inline autogluon/training/docker/0.2/py3/Dockerfile.cpu:135]: @jeet4320 should this be included before the oss compliance script?

File: autogluon/training/docker/0.2/py3/cu102/Dockerfile.gpu [hunk 17]
```diff
@@ -0,0 +1,85 @@
...
+ARG PYTHON_VER=3.7
+ARG RAPIDS_VER=0.19
+ARG FROM_IMAGE=rapidsai/rapidsai-core
+
+FROM ${FROM_IMAGE}:${RAPIDS_VER}-cuda${CUDA_VER}-base-${LINUX_VER}-py${PYTHON_VER}
+
+LABEL maintainer="Amazon AI"
+LABEL dlc_major_version="1"
+
+ARG CUDA_VER
```
Comment (tejaschumbalkar) [inline autogluon/training/docker/0.2/py3/cu102/Dockerfile.gpu:13]: nit: `ARG CUDA_VER=10.2` is defined on line 1

File: test/sagemaker_tests/autogluon/training/integration/sagemaker/test_training.py [hunk 18]
```diff
@@ -0,0 +1,60 @@
...
+        instance_count=1,
+        instance_type=instance_type,
+        sagemaker_session=sagemaker_session,
+        image_uri=ecr_image,
+        framework_version=framework_version,
+    )
+
+    ag = _disable_sm_profiler(sagemaker_session.boto_region_name, ag)
+
+    with timeout(minutes=15):
```
Comment (tejaschumbalkar) [inline test/sagemaker_tests/autogluon/training/integration/sagemaker/test_training.py:41]: nit: Can we use `DEFAULT_TIMEOUT` variable defined in `__init__.py` ?

File: test/testrunner.py [hunk 19]
```diff
@@ -316,6 +316,14 @@ def main():
         sm_utils.generate_empty_report(report, test_type, "huggingface")
         return
 
+    # Skipping non AutoGluon specific tests to execute only sagemaker tests
```
Comment (tejaschumbalkar) [inline test/testrunner.py:319]: Seems that this logic is similar to the one above used for `huggingface`. Either configure the above to be used for autogluon or define the common logic in a separate function and call it here.

File: autogluon/training/docker/0.2/py3/cu102/Dockerfile.gpu (hunk 17 above)
Comment (SergTogul) [reply autogluon/training/docker/0.2/py3/cu102/Dockerfile.gpu:13]: https://github.com/aws/deep-learning-containers/pull/1202#discussion_r676941928

File: autogluon/training/docker/0.2/py3/Dockerfile.cpu [hunk 20]
```diff
@@ -0,0 +1,137 @@
...
+  sed 's@session\s*required\s*pam_loginuid.so@session optional pam_loginuid.so@g' -i /etc/pam.d/sshd
+
+RUN rm -rf /root/.ssh/ \
+ && mkdir -p /root/.ssh/ \
+ && ssh-keygen -q -t rsa -N '' -f /root/.ssh/id_rsa \
+ && cp /root/.ssh/id_rsa.pub /root/.ssh/authorized_keys \
+ && printf "Host *\n StrictHostKeyChecking no\n" >> /root/.ssh/config
+
+# This is here to make our installed version of OpenCV work.
+# https://stackoverflow.com/questions/29274638/opencv-libdc1394-error-failed-to-initialize-libdc1394
```
[the remark this answers is not in the document]
Comment (SergTogul) [reply autogluon/training/docker/0.2/py3/Dockerfile.cpu:122]: #1261

File: autogluon/training/docker/0.2/py3/Dockerfile.cpu [hunk 21]
```diff
@@ -0,0 +1,137 @@
...
+ && pip --no-cache-dir install --upgrade wheel setuptools
+
+WORKDIR /
+
+RUN pip install --no-cache-dir -U \
+    torch==${TORCH_VER} torchvision==${TORCH_VISION_VER} -f https://download.pytorch.org/whl/torch_stable.html \
+    mxnet==${MXNET_VER} \
+    autogluon==${AUTOGLUON_VERSION} \
+    sagemaker-mxnet-training \
+&& pip uninstall -y jupyter-client jupyter-core jupyter-packaging jupyter-server jupyter-server-proxy \
```
[the remark this answers is not in the document]
Comment (SergTogul) [reply autogluon/training/docker/0.2/py3/Dockerfile.cpu:101]: Saving space

**Pull request merged**